connector/mqtt/vernemq/v2k-brigde/app/MqttClient.js

```diff
@@ -1,68 +1,128 @@
+    this.keepAlive = this.config.mqtt.keepAlive;
+    this.port = this.config.mqtt.port;
     this.username = this.config.mqtt.clientUsername;
+    this.secureMode = this.config.mqtt.secure;
+
+    this.privateKey = fs.readFileSync(`${this.config.mqtt.tls.privateKey.location}`);
+    this.clientCrt = fs.readFileSync(`${this.config.mqtt.tls.certificate.location}`);
+    this.ca = fs.readFileSync(`${this.config.mqtt.tls.ca.location}`);
+
+    // backPressure
+    this.messageQueue = null;
+    this.currentMessageQueueLenght = 0;
 
-    this.key = fs.readFileSync(path.join(__dirname, `${this.config.mqtt.hostname}.key`));
-    this.clientCrt = fs.readFileSync(path.join(__dirname, `${this.config.mqtt.hostname}.crt`));
-    this.caCrt = fs.readFileSync(path.join(__dirname, 'ca.crt'));
+    // agent messenger
+    this.agentMessenger = agentMessenger;
   }
 
   init() {
-    const mqttOptions = {
+    this.mqttOptions = {
       username: this.username,
       clientId: this.clientId,
-      host: this.config.mqtt.host,
-      port: this.config.mqtt.port,
-      protocol: 'mqtts',
-      ca: this.caCrt,
-      key: this.key,
+      host: this.host,
+      port: this.port,
+      protocol: this.secureMode ? 'mqtts' : 'mqtt',
+      ca: this.ca,
+      key: this.privateKey,
       cert: this.clientCrt,
-      keepAlive: this.config.mqtt.keepAlive,
+      keepAlive: this.keepAlive,
       clean: false,
-      rejectUnauthorized: true,
+      rejectUnauthorized: false,
     };
 
     const onConnectBind = this.onConnect.bind(this);
     const onDisconnectBind = this.onDisconnect.bind(this);
     const onMessageBind = this.onMessage.bind(this);
 
-    this.mqttc = mqtt.connect(mqttOptions);
+    this.connect();
     this.mqttc.on('connect', onConnectBind);
     this.mqttc.on('disconnect', onDisconnectBind);
     this.mqttc.on('message', onMessageBind);
+
+    /**
+     * Create async queue
+     */
+    this.messageQueue = async.queue((data, done) => {
+      this.asyncQueueWorker(data);
+      done();
+    }, PARALLEL_HANDLERS);
+
+    /**
+     * When processing was finalized, reconnect to broker
+     */
+    this.messageQueue.drain(() => {
+      if (this.isConnected === false) {
+        this.mqttc.reconnect();
+      }
+    });
   }
 
   onConnect() {
     this.isConnected = true;
     logger.info(`Client ${this.clientId} connected successfully!`, TAG);
-
-    if (this.agentMessenger === null) {
-      this.agentMessenger = new AgentMessenger(this.config);
-      this.agentMessenger.init(this.mqttc);
-    }
+    this.subscribe();
   }
 
   onDisconnect() {
+    logger.info(`Client ${this.clientId} disconnected, reconnecting ......`);
     this.isConnected = false;
     this.mqttc.reconnect();
   }
 
   onMessage(topic, message) {
+    // pause
+    if (this.currentMessageQueueLenght > MAX_QUEUE_LENGTH) {
+      this.mqttc.end(true);
+      this.isConnected = false;
+      return;
+    }
+
+    if (this.isConnected) {
```

Comment:
When I've been testing I noticed duplicated messages, after a connection event to the broker because of the clean session.
I found a way to check if the message is duplicated, in the packet received on the message callback.
Thanks for the review.